packages/tizen-remote/lib/index.js

```diff
@@ -0,0 +1,742 @@
+    if (this.#ws) {
+      const listeners = this.#listeners.get(event) ?? new Set();
+      this.#listeners.set(event, listeners);
+      if (context) {
+        const boundListener = /** @type {Listener} */(listener.bind(context));
+        listeners.add(boundListener);
+        this.#ws.on(event, boundListener);
+        return boundListener;
+      }
+      listeners.add(listener);
+      this.#ws.on(event, listener);
+      return listener;
+    }
+    throw new Error('Not connected');
+  }
+
+  /**
+   *
+   * @template {(...args: any[]) => void} Listener
+   * @param {WsEvent} event
+   * @param {Listener} listener
+   * @param {{ context?: any }} [opts]
+   * @returns {Listener}
+   */
+  #onceWs(event, listener, {context} = {}) {
+    if (this.#ws) {
+      const listeners = this.#listeners.get(event) ?? new Set();
+      this.#listeners.set(event, listeners);
+      if (context) {
+        const boundListener = /** @type {Listener} */(listener.bind(context));
+        listeners.add(boundListener);
+        this.#ws.once(event, boundListener);
+        return boundListener;
+      }
+      listeners.add(listener);
+      this.#ws.once(event, listener);
+      return listener;
+    }
+    throw new Error('Not connected');
+  }
+
+  /**
+   *
+   * @template {(...args: any[]) => void} Listener
+   * @param {WsEvent} event
+   * @param {Listener} listener
+   * @returns {void}
+   */
+  #offWs(event, listener) {
+    if (this.#ws) {
+      let foundListener = Boolean(this.#listeners.get(event)?.has(listener));
+      if (foundListener) {
+        this.#ws.removeListener(event, listener);
+        this.#listeners.get(event)?.delete(listener);
+      }
+    }
+  }
+
+  /**
+   * Execute a "click" on the remote
+   * @param {KeyCode} key
+   */
+  async click(key) {
+    await this.send(new KeyCommand(KeyCmd.CLICK, key));
+  }
+
+  /**
+   * Execute a "press" ("keydown") on the remote
+   * @param {KeyCode} key
+   */
+  async press(key) {
+    await this.send(new KeyCommand(KeyCmd.PRESS, key));
+  }
+
+  /**
+   * Execute a "release" ("keyup") on the remote
+   * @param {KeyCode} key
+   */
+  async release(key) {
+    await this.send(new KeyCommand(KeyCmd.RELEASE, key));
+  }
+
+  /**
+   * Execute a "long" press on the remote
+   * @param {KeyCode} key
+   */
+  async longPress(key, ms = 1000) {
+    await this.send(new KeyCommand(KeyCmd.PRESS, key));
+    await delay(ms);
+    await this.send(new KeyCommand(KeyCmd.RELEASE, key));
+  }
+
+  /**
+   * Gets a new token from the Tizen device (if none exists).
+   *
+   * If a new token must be requested, expect to wait _at least_ thirty (30) seconds.
+   * @returns {Promise<string>}
+   */
+  async getToken() {
+    if (this.#token) {
+      return this.#token;
+    }
+    const res = await this.sendRequest(
+      constants.TOKEN_EVENT,
+      new KeyCommand(KeyCmd.CLICK, Keys.HOME)
+    );
+    if (res?.data?.token) {
+      this.emit(Event.TOKEN, res.data.token);
```

Comment:
why bother emitting instead of just assigning `this.#token` here?

Reply:
nothing is listening for `Event.TOKEN`, so it's just for consumers (especially since it takes a long time!).  But I only want to assign the value of `this.#token` if a `TizenRemote` instance called `getToken()`, which is why we just return the value here.  The assignment happens in `connect()` after a successful call to `getToken()`.